src/models.py

```diff
@@ -171,6 +182,63 @@ def relu_attn(self, query, key, value, attention_mask=None, head_mask=None):
 
     return attn_output, attn_weights
 
+# used in nystrom_attn
+# iteratively finds the Moore-Penrose inverse
+# work-in-progress
+def pinv(a_s):
+    abs_x = torch.abs(a_s)
+    col = abs_x.sum(dim = -1)
+    row = abs_x.sum(dim = -2)
+
+    # Need to double check this initialization, doesn't seem to have right size
+    current_z = a_s.transpose(-1, -2) / (torch.max(col) * torch.max(row))
+    next_z = pinv_next(a_s, current_z)
+
+    while(not torch.eq(current_z, next_z)):
+        current_z = next_z
+        next_z = pinv_next(a_s, current_z)
+    return next_z
+
+# helper function for pinv
+def pinv_next(a_s, z):
+    dim = a_s.shape[0]
+    i = torch.eye(dim)
+    inner = 7*i-torch.matmul(a_s,z)
+    inner = 15 * i - torch.matmul(torch.matmul(a_s, z), inner)
+    inner = 13 * i - torch.matmul(torch.matmul(a_s, z), inner)
+    return 1/4*torch.matmul(z, inner)
+
+# needed for landmarking in nystrom, downsamples
+def segmented_means(matrix, m = 2):
+    dim3 = matrix.shape[2] #this is the dim we downsample
+    pooler = nn.AvgPool2d((math.ceil(dim3/m), 1), ceil_mode = True)
+    return pooler(matrix)
+
+# m must be much smaller than dimension, we should probably decide on an m
+def nystrom_attn(self, query, key, value, attention_mask=None, head_mask=None, m = 2):
+    # landmarking
+    q_bar = segmented_means(query, m)#query[:,:,:m]
+    k_bar = segmented_means(key, m)#key[:,:,:m]
+
+    # this was taken directly from the nystrom paper
+    einops_eq = '... i d, ... j d -> ... i j'
+    sim1 = einsum(einops_eq, query, k_bar)
+    sim2 = einsum(einops_eq, q_bar, k_bar)
+    sim3 = einsum(einops_eq, q_bar, key)
+
+    # this was also taken from the nystrom paper, but modified
+    attn1, attn2, attn3 = map(lambda t: t.softmax(dim = -1), (sim1, sim2, sim3))
+    attn2_inv = torch.linalg.pinv(attn2)
```

Comment:
Reply to this message with the speed measurements -- basically what's the speedup of `torch.linalg.pinv` over the nystrom `pinv` for varying values of `m`? (do just like 2-3 runs for m=2,5,8)

Reply:
### training time using 51 steps:
**pytorch method:**
m=8: 1:13
m=5: 1:11
m=2: 1:07
**iterative method with 7 iterations:**
m=8: 1:21
m=5: 1:15
m=2: 1:08